packages/flutter/lib/src/services/text_boundary.dart

```diff
@@ -0,0 +1,179 @@
+      return const TextPosition(offset: 0);
+    }
+    if (position.offset > _text.length ||
+        (position.offset == _text.length && position.affinity == TextAffinity.downstream)) {
+      return TextPosition(offset: _text.length, affinity: TextAffinity.upstream);
+    }
+    final int endOffset;
+    final int startOffset;
+    switch (position.affinity) {
+      case TextAffinity.upstream:
+        startOffset = math.min(position.offset - 1, _text.length);
+        endOffset = math.min(position.offset, _text.length);
+        break;
+      case TextAffinity.downstream:
+        startOffset = math.min(position.offset, _text.length);
+        endOffset = math.min(position.offset + 1, _text.length);
+        break;
+    }
+    return TextPosition(
+      offset: CharacterRange.at(_text, startOffset, endOffset).stringBeforeLength,
+    );
+  }
+
+  @override
+  TextPosition getTrailingTextBoundaryAt(TextPosition position) {
+    if (position.offset < 0 ||
+        (position.offset == 0 && position.affinity == TextAffinity.upstream)) {
+      return const TextPosition(offset: 0);
+    }
+    if (position.offset >= _text.length) {
+      return TextPosition(offset: _text.length, affinity: TextAffinity.upstream);
+    }
+    final int endOffset;
+    final int startOffset;
+    switch (position.affinity) {
+      case TextAffinity.upstream:
+        startOffset = math.min(position.offset - 1, _text.length);
+        endOffset = math.min(position.offset, _text.length);
+        break;
+      case TextAffinity.downstream:
+        startOffset = math.min(position.offset, _text.length);
+        endOffset = math.min(position.offset + 1, _text.length);
+        break;
+    }
+    final CharacterRange range = CharacterRange.at(_text, startOffset, endOffset);
+    return TextPosition(
+      offset: _text.length - range.stringAfterLength,
+      affinity: TextAffinity.upstream,
+    );
+  }
+}
+
+/// A text boundary that uses words as logical boundaries.
+///
+/// This class uses [UAX #29](https://unicode.org/reports/tr29/) defined word
+/// boundaries to calculate its logical boundaries.
+class WordBoundary extends TextBoundary {
+  /// Creates a [CharacterBoundary] with the text and layout information.
+  const WordBoundary(this._textLayout);
+
+  final TextLayoutMetrics _textLayout;
+
+  @override
+  TextPosition getLeadingTextBoundaryAt(TextPosition position) {
+    return TextPosition(
+      offset: _textLayout.getWordBoundary(position).start,
+      // Word boundary seems to always report downstream on many platforms.
+      affinity: TextAffinity.downstream,  // ignore: avoid_redundant_argument_values
+    );
+  }
+  @override
+  TextPosition getTrailingTextBoundaryAt(TextPosition position) {
+    return TextPosition(
+      offset: _textLayout.getWordBoundary(position).end,
+      // Word boundary seems to always report downstream on many platforms.
+      affinity: TextAffinity.downstream,  // ignore: avoid_redundant_argument_values
```

Comment:
shouldn't this be `upstream`, following `LineBreak`'s logic? Or should the word breaker behave differently?

Reply:
nice catch, will fix